### PR TITLE
audit functionalities added

### PR DESCRIPTION
**Pull request by yipingbenjaminliu**
#302 
- small clarification needed for comparing rolls for attendance monitors
- adjusted trainee population script to more accurately represent users and more robust
- small known bug that requires pm clarification
- wrote a roll population script

### CODE REVIEW

File: ap/attendance/views.py [hunk 1]
```diff
@@ -157,6 +161,131 @@ def get_context_data(self, **kwargs):
...
+    
+    ctx['current_period'] = Term.period_from_date(Term.current_term(), date.today())
+    period_begin = Term.period_from_date(Term.current_term(), date.today())
+    period_end= Term.period_from_date(Term.current_term(), date.today())
+
+    trainees_secondyear = Trainee.objects.filter(current_term__gt=2, gender="B").exclude(groups__name="attendance_monitors")
+
+    if self.request.method == 'GET':
+      gen = self.request.GET.get('gender')
+      if gen == "B":
```
Comment (radicalray) [inline ap/attendance/views.py]: I bet you can do something like this here:
```
trainees_secondyear = Trainee.objects.filter(current_term_gt=2)

if gen == "S" or gen == "B":
  trainees_secondyear.filter(gender=gen)
elif "B" in gen and "S" in gen:
  # Do nothing
else:
  trainees_secondyear = Trainee.objects.none()

```

File: ap/attendance/views.py [hunk 2]
```diff
@@ -157,6 +161,131 @@ def get_context_data(self, **kwargs):
...
+        trainees_secondyear = Trainee.objects.filter(current_term__gt=2, gender="B")
+      elif gen == "S":
+        trainees_secondyear = Trainee.objects.filter(current_term__gt=2, gender="S")
+      elif (gen == "BS" or gen == "SB"):
+        trainees_secondyear = Trainee.objects.filter(current_term__gt=2)
+      elif (gen == ""):
+        trainees_secondyear = Trainee.objects.none()
+      
+      per = str(self.request.GET.get('period'))
```
Comment (radicalray) [inline ap/attendance/views.py]: if per is None, that's the same as nil in ruby or undefined/null in javascript. You don't have to cast to string. You can simply get per

```
per = self.request.GET.get('period')

if per:
  if per <= 9:
    period_begin = period_end = per
  else:
    per_str = str(per)
    period_being, period_end = int(per_str[0]), int(per_str[-1])
else:
  period_begin = period_end = 10
```

Comment (radicalray) [inline ap/attendance/views.py]: On second thought, you shouldn't create a period range like this, you'll run into a lot of trouble. Make it more explicit. Just pass in 2 get arguments: self.request.GET.get('period_start') and self.request.GET.get('period_end'). If you MUST squish it all into 1 variable, please always put a separator: per = '1-9' or per='1:9' so you can do a per.split('-') to get out the start/end more clearly.

File: ap/attendance/views.py [hunk 3]
```diff
@@ -157,6 +161,131 @@ def get_context_data(self, **kwargs):
...
+      AT_discrepancy = 0
+      details = ""
+      rolls = Roll.objects.filter(trainee=t, date__gte=start_date, date__lte=end_date)
+      roll_trainee = rolls.filter(submitted_by=t) #rolls taken by trainee
+      roll_am = rolls.filter(submitted_by=am) #rolls taken by attendance monitor
+      for r in roll_am:
+        r_stat_trainee = roll_trainee.filter(event=r.event, date=r.date).values('status')[0] #status of correspond event from trainee        
+        r_stat_trainee = r_stat_trainee['status']
+
+        # PM indicates that mismatch is only when trainee marks P and AM marks otherwise
```
Comment (radicalray) [inline ap/attendance/views.py]: What about the other way? trainee marks tardy but AM marks absent or trainee marks present and AM marks tardy?

File: ap/attendance/views.py [hunk 4]
```diff
@@ -157,6 +161,131 @@ def get_context_data(self, **kwargs):
...
+      rolls = Roll.objects.filter(trainee=t, date__gte=start_date, date__lte=end_date)
+      roll_trainee = rolls.filter(submitted_by=t) #rolls taken by trainee
+      roll_am = rolls.filter(submitted_by=am) #rolls taken by attendance monitor
+      for r in roll_am:
+        r_stat_trainee = roll_trainee.filter(event=r.event, date=r.date).values('status')[0] #status of correspond event from trainee        
+        r_stat_trainee = r_stat_trainee['status']
+
+        # PM indicates that mismatch is only when trainee marks P and AM marks otherwise 
+        if r_stat_trainee == 'P' and r.status != 'P':
+          mismatch = mismatch+1
```
Comment (radicalray) [inline ap/attendance/views.py]: You can do shorthand ^
mismatch += 1

File: ap/attendance/views.py [hunk 5]
```diff
@@ -157,6 +161,131 @@ def get_context_data(self, **kwargs):
...
+          details = details + ' MF' + str(r.date.month) + '/' + str(r.date.day) + " " + r.event.code + ","
```
Comment (radicalray) [inline ap/attendance/views.py]: string interpolation reads better:
```
details = []
...
details.push("MF %d/%d %s" % (r.date.month, r.date.day, r.event.code))
...
details.join(', ')
```

File: ap/attendance/views.py [hunk 6]
```diff
@@ -157,6 +161,131 @@ def get_context_data(self, **kwargs):
...
+
+        # #PM indicates that AT discrepancy is only when AM marks A and trainee marks a type of T
+        if r.status == 'A' and (r_stat_trainee == 'T' or r_stat_trainee == 'U' or r_stat_trainee == 'L'):
```
Comment (radicalray) [inline ap/attendance/views.py]: @billyeh used this shorthand:
```
r.status == 'A' and r_stat_trainee in set(['T', 'U', 'L'])
```

File: ap/attendance/views.py [hunk 7]
```diff
@@ -157,6 +161,131 @@ def get_context_data(self, **kwargs):
...
+          AT_discrepancy = AT_discrepancy +1
+          details = details + ' AT' + str(r.date.month) + '/' + str(r.date.day) + " " + r.event.code + ","
```
Comment (radicalray) [inline ap/attendance/views.py]: Same comments here as above ^

File: ap/attendance/views.py [hunk 8]
```diff
@@ -157,6 +161,131 @@ def get_context_data(self, **kwargs):
...
+  def post(self, request, *args, **kwargs):
+    context = self.get_context_data()
+    return super(TrackAuditList, self).render_to_response(context)
+  
+  def get_context_data(self, **kwargs):
+    ctx = super(TrackAuditList, self).get_context_data(**kwargs)
+    #2nd year trainees under audit list have self_attendace set as falst
+
+    if self.request.method == 'POST':
+      aud = self.request.POST.get('type')
```
Comment (radicalray) [inline ap/attendance/views.py]: Better naming convention. This is not clear. ```type``` is the type of action? and add/del is giving trainees ability to take self attendance? If so, how about rename type to "self_attendance" and the value to "on"/"off"?

File: ap/meal_seating/models.py [hunk 9]
```diff
@@ -22,7 +22,7 @@ class Table(models.Model):
     )
 
     location = models.CharField(max_length=2, choices=LOCATIONS)
-    genderType = models.CharField(max_length=1, choices=GENDERS)
+    genderType = models.CharField(max_length=1, choices=GENDERS, default='S')
```
Comment (radicalray) [inline ap/meal_seating/models.py]: Why is the gender Sister by default?

File: ap/services/models/assignment.py [hunk 10]
```diff
@@ -13,7 +13,7 @@ class Assignment(models.Model):
 
     service = models.ForeignKey('Service', related_name='assignments')
     # Get role + workload
-    pool = models.ForeignKey('AssignmentPool')
+    pool = models.ForeignKey('AssignmentPool', default='null')
```
Comment (radicalray) [inline ap/services/models/assignment.py]: Why is this modified for an audit pr? Please resolve migration issues independent of model code changes.

**Comment by yipingbenjaminliu** [comment]
Thanks @radicalray, this is what I needed!

File: ap/meal_seating/models.py (hunk 9 above)
Comment (yipingbenjaminliu) [inline ap/meal_seating/models.py]: this was an flag raised by running migrate, it required a default value. I thought since there are more sisters than brothers, then default to S.

File: ap/services/models/assignment.py (hunk 10 above)
Comment (yipingbenjaminliu) [inline ap/services/models/assignment.py]: this was an flag raised by running migrate, it required a default value and I set this just to run migrate through. I'll take this out.

**Comment by yipingbenjaminliu** [comment]
I was running into some issues with running the population scripts so I ran migrate and makemigrations but encountered some issues from the system asking for default values. If I run it right now I'll still encounter some issues with the database due to the lack of default values. Maybe someone should address this down the line for database robustness.

**Comment by radicalray** [comment]
@yipingbenjaminliu Yea, migrations are still very unstable within django. This is a known difficult problem. Do whatever is most expedient to get it to work on your local dev machine but do keep the code as clean as possible for the sake of the rest of the rest of the team. You can always blow everything away and pull in production migrations like via this wiki page: https://github.com/attendanceproject/djattendance/wiki/Makemigrations-notes

If you find better ways of doing this, feel free to add to it.

File: ap/ap/settings/base.py [hunk 11]
```diff
@@ -235,26 +235,23 @@
 # more details on how to customize your logging configuration.
 LOGGING = {
     'version': 1,
-    'disable_existing_loggers': False,
-    'filters': {
-        'require_debug_false': {
-            '()': 'django.utils.log.RequireDebugFalse'
-        }
-    },
+    'disable_existing_loggers': True,
```
Comment (radicalray) [inline ap/ap/settings/base.py]: What's the main motivation behind deleting the filters? Is this something we want in production and other's dev environments?

File: ap/attendance/templates/attendance/roll_audit.html [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+
+$(document).ready(function() {
+  if (gen == "B"){
+    $("#brothers").button('toggle')
+  }else if (gen == "S"){
+    $("#sisters").button('toggle')
+  };
+  
+  for(var i = per; i<=9; i++){
+    $("#"+i).attr('disabled', true);
```
Comment (radicalray) [inline ap/attendance/templates/attendance/roll_audit.html]: Any reason why this isn't done in the django template? You can use ```counter```. See: http://stackoverflow.com/questions/11481499/django-iterate-number-in-for-loop-of-a-template

File: ap/ap/settings/base.py [hunk 13]
```diff
@@ -235,12 +235,12 @@
...
-    'disable_existing_loggers': False,
-    'filters': {
-        'require_debug_false': {
-            '()': 'django.utils.log.RequireDebugFalse'
-        }
-    },
+     'disable_existing_loggers': False,
+     'filters': {
+         'require_debug_false': {
+             '()': 'django.utils.log.RequireDebugFalse'
```
Comment (radicalray) [inline ap/ap/settings/base.py]: I think there's an extra space in front of each line.

File: ap/attendance/management/commands/populate_rolls.py [hunk 14]
```diff
@@ -0,0 +1,40 @@
+from django.core.management.base import BaseCommand
+from attendance.models import Roll
+from schedules.models import Event
+from accounts.models import Trainee
+from terms.models import Term
+import random
+
+def new_roll(trainees=[], am=[]):
```
Comment (radicalray) [inline ap/attendance/management/commands/populate_rolls.py]: Once all these changes get merged in, can you update our wiki on how to popular rolls and stuff?

https://github.com/attendanceproject/djattendance/wiki/Generating-Dummy-Data

thanks!

File: ap/attendance/views.py [hunk 15]
```diff
@@ -157,6 +161,121 @@ def get_context_data(self, **kwargs):
...
+        trainees_secondyear = Trainee.objects.none()
+      
+      per = self.request.GET.get('period')
+
+      if per:
+        if len(per) == 1:
+          period_begin = int(per)
+          period_end = int(per)
+        elif len(per) > 1 :
+          period_begin = int(per[0])
```
Comment (radicalray) [inline ap/attendance/views.py]: This could break if we ever have more than 9 periods. Can you add this as a comment for future devs to know? thanks

File: ap/attendance/templates/attendance/audit_track.html [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+
+
+$(document).ready(function() {
+  $("#addtrainee").select2({
+    placeholder: "Select a trainee",
+    allowClear: true
+  }).on("select2:select", function(e) {
+    audit($('#addtrainee').val(), "no")
+  });
+  $( "body" ).on('click', '.btn', function(){
```
Comment (radicalray) [inline ap/attendance/templates/attendance/audit_track.html]: This will bind every .btn class on the page, could be dangerous someone adds a new button to the page in the future. Please scope the selector more. ```.audit-btn``` or ```.audit .btn``` or something

File: ap/attendance/templates/attendance/roll_audit.html [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+  <tr>
+    <td>Show:&nbsp;
+      <div class="btn-group" role="group">  
+        <button type="button" class="btn btn-primary" data-toggle="button" id="brothers">brothers</button>
+        <button type="button" class="btn btn-primary" data-toggle="button" id="sisters">sisters</button>&nbsp;&nbsp;
+      </div></td>
+      <td>  Periods:   
+        <div class="btn-group" role="group">
+          {% for i in "123456789" %}
+            <button type="button" class="btn btn-primary" data-toggle="button" id="{{ i }}">{{ i }}</button>
```
Comment (radicalray) [inline ap/attendance/templates/attendance/roll_audit.html]: You should just print the disabled here:

```
<button type="button" class="btn btn-primary" data-toggle="button" id="{{ i }}" {% if i > current_period %} disabled {% end %|>{{ i }}</button>
```

instead of using javascript.

File: ap/attendance/templates/attendance/roll_audit.html [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+$(document).ready(function() {
+  if (gen == "B"){
+    $("#brothers").button('toggle')
+  }else if (gen == "S"){
+    $("#sisters").button('toggle')
+  };
+
+  {% for i in "123456789" %}
+    {% if i < per %}
+      $("#"+i).attr('disabled', true);
```
Comment (radicalray) [inline ap/attendance/templates/attendance/roll_audit.html]: This is definitely not what I meant...see comment below.

File: ap/attendance/templates/attendance/roll_audit.html [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+    {% endif %}
+  {% endfor %}
+
+  $( "body" ).on('click', '.btn-info', function(){
```
Comment (radicalray) [inline ap/attendance/templates/attendance/roll_audit.html]: Try to scope the selectors here and below more to avoid future collisions.

File: ap/attendance/templates/attendance/roll_audit.html (hunk 17 above)
Comment (radicalray) [inline ap/attendance/templates/attendance/roll_audit.html]: Also, avoid having just an integer as your id, may not work across browsers. Something like ```id="per-{{ i }}"``` would be preferred.

File: ap/attendance/views.py [hunk 20]
```diff
@@ -157,6 +161,121 @@ def get_context_data(self, **kwargs):
...
+      details = []
+      if period_begin:
+        rolls = Roll.objects.filter(trainee=t, date__gte=start_date, date__lte=end_date)
+        roll_trainee = rolls.filter(submitted_by=t) #rolls taken by trainee
+        roll_am = rolls.filter(submitted_by=am) #rolls taken by attendance monitor
+        for r in roll_am:
+          r_stat_trainee = roll_trainee.filter(event=r.event, date=r.date).values('status')[0] #status of correspond event from trainee        
+          r_stat_trainee = r_stat_trainee['status']
+
+          # PM indicates that mismatch is only when trainee marks P and AM marks otherwise
```
Comment (radicalray) [inline ap/attendance/views.py]: Can you crystallize the business logic here for determining a discrepancy as a block of comments on the top of the file so if that changes in the future, it's easier to interpret? thanks!

**Comment by radicalray** [comment]
demo

**Comment by radicalray** [comment]
Don't forget to pull in the latest dev and rebase. It seems like you have some conflicts you need to resolve before merging.

File: ap/attendance/management/commands/populate_rolls.py (hunk 14 above)
Comment (yipingbenjaminliu) [inline ap/attendance/management/commands/populate_rolls.py]: complete

File: ap/attendance/views.py (hunk 15 above)
Comment (yipingbenjaminliu) [inline ap/attendance/views.py]: complete

File: ap/attendance/views.py (hunk 20 above)
Comment (yipingbenjaminliu) [inline ap/attendance/views.py]: complete

**Comment by yipingbenjaminliu** [comment]
Still need to rebase from dev and create the gif file for UI operation

File: ap/attendance/views.py [hunk 21]
```diff
@@ -158,6 +162,128 @@ def get_context_data(self, **kwargs):
...
+  
+  def get_context_data(self, **kwargs):
+    ctx = super(AuditRollsView, self).get_context_data(**kwargs)
+    ctx['current_url'] = resolve(self.request.path_info).url_name
+    ctx['user_gender'] = Trainee.objects.filter(id=self.request.user.id).values('gender')[0]
+    ctx['current_period'] = Term.period_from_date(Term.current_term(), date.today())
+
+    audit_log = []
+    period_begin = Term.period_from_date(Term.current_term(), date.today())
+    period_end= Term.period_from_date(Term.current_term(), date.today())
```
Comment (jontey) [inline ap/attendance/views.py]: These two lines return the same thing. Are you planing on supporting a date range in the future?

Comment (yipingbenjaminliu) [inline ap/attendance/views.py]: they're set as a placeholder for the default load, the input values are within an if clause and the default input is null, this is placed so the load default doesn't throw errors. there's a better way to do this, i just don't know how

File: ap/attendance/templates/attendance/roll_audit.html [hunk 22]
```diff
@@ -1,170 +1,242 @@
...
-  if (gen == "B"){
-    $("#brothers").button('toggle')
-  }else if (gen == "S"){
-    $("#sisters").button('toggle')
-  };
 
-  // for(var i = per; i <= 9; i++){
-  //   $("#per"+i).attr('disabled', true)
-  // };
+  $("body").on('click', 'button.btn', function() {
```
Comment (jontey) [inline ap/attendance/templates/attendance/roll_audit.html]: jquery has toggleClass http://api.jquery.com/toggleclass/

File: ap/attendance/views.py [hunk 23]
```diff
@@ -166,6 +171,85 @@ def get_context_data(self, **kwargs):
 
     return ctx
 
+<<<<<<< HEAD
```
Comment (jontey) [inline ap/attendance/views.py]: Remove this
